src/libfsm/union.c

```diff
@@ -157,231 +208,874 @@ fsm_union_array(size_t fsm_count,
+		edge_set_group_iter_reset(ns->edges, EDGE_GROUP_ITER_ALL, &iter);
+		while (edge_set_group_iter_next(&iter, &info)) {
+			assert(info.to < nfa->statecount);
+			const struct state_set *to_eclosure = eclosures[info.to];
+
+			struct state_iter dst_si;
+			state_set_reset(to_eclosure, &dst_si);
+			fsm_state_t dst_s_i;
+			while (state_set_next(&dst_si, &dst_s_i)) {
+				if (dst_s_i == unanchored_end_loop) {
+					if (info.to != unanchored_end_loop) {
+						*indirect_dst = info.to;
+					}
+
+					return true;
+				}
+			}
+		}
+	}
+
+	return false;
+}
+
+static bool
+start_state_epsilon_closure_matches_empty_string(const struct fsm *nfa, const struct state_set *eclosure)
+{
+	struct state_iter si;
+	state_set_reset(eclosure, &si);
+
+	fsm_state_t s_i;
+	while (state_set_next(&si, &s_i)) {
+		if (fsm_isend(nfa, s_i)) { return true; }
+	}
+
+	return false;
+}
+
+static const struct fsm_options dump_nfa_opt = {
+	.io = FSM_IO_STR,
+	.ambig = AMBIG_MULTIPLE,
+	.case_ranges = 1,
+	.consolidate_edges = 1,
+	.group_edges = 1,
+};
+
+static bool
+analyze_group_nfa(const struct fsm *nfa, struct analysis_info *ainfo)
+{
+	if (LOG_ANALYZE_GROUP_NFA) {
+		fprintf(stderr, "==== %s\n", __func__);
+		if (LOG_ANALYZE_GROUP_NFA > 1) {
+			fsm_print(stderr, nfa, &dump_nfa_opt, NULL, FSM_PRINT_DOT);
+		}
+	}
+
+	memset(ainfo, 0x00, sizeof(*ainfo));
+	ainfo->start = NO_STATE;
+	ainfo->unanchored_start_loop = NO_STATE;
+	ainfo->anchored_start = NO_STATE;
+	ainfo->unanchored_end_loop = NO_STATE;
+	ainfo->unanchored_end_loop_end = NO_STATE;
+	ainfo->eager_match_state = NO_STATE;
+
+	if (!fsm_getstart(nfa, &ainfo->start)) {
+		return false;
+	}
+
+	const size_t state_count = fsm_countstates(nfa);
+	assert(ainfo->start < state_count);
+
+	struct state_set **eclosures = epsilon_closure(nfa);
+	if (eclosures == NULL) {
+		return false;
+	}
+
+	/* First pass: Iterate over the start state's epsilon edges,
+	 * attempting to identify the unanchored start loop and anchored
+	 * start states (if present).
+	 *
+	 * Note: This uses the start state's epsilon set rather than its
+	 * epsilon closure because (by construction) the unanchored
+	 * start loop and anchored start states will both be directly
+	 * connected to the start state. Using the epsilon closure can
+	 * mis-identify the unanchored *end* loop as the start loop, if
+	 * there is a path with only epsilon edges between them. */
+	struct state_iter si;
+	state_set_reset(nfa->states[ainfo->start].epsilons, &si);
+	fsm_state_t ns_i;
+	while (state_set_next(&si, &ns_i)) {
+		if (ns_i == ainfo->start) { continue; }
+
+		/* If there's a state in the start state's epsilon set that
+		 * has a dot self-edge, it's the unanchored start loop. */
+		if (has_dot_self_edge(nfa, ns_i)) {
+			if (LOG_ANALYZE_GROUP_NFA) {
+				fprintf(stderr, "%s: unanchored_start_loop found on state %d\n", __func__, ns_i);
+			}
+
+			/* TODO: There is only one unanchored start loop, but in obscure cases it may
```

Comment:
Basically, fuzzing found a pathological case here that depends on a regex with a `\0` in the middle of it. I can't reproduce it otherwise, and it's not present in our input, so I'm just leaving it as a failing-in-fuzzing assert to investigate later.

Same with the other assert about 20 lines below.